feat(files_sharing): Migrate public shares to Vue

susnux opened the pull request:
* Resolves: https://github.com/nextcloud/server/issues/42968
* Resolves: https://github.com/nextcloud/server/issues/4109
  * Resolves: https://github.com/nextcloud/server/issues/673
* Resolves: https://github.com/nextcloud/server/issues/45510
* Resolves: https://github.com/nextcloud/server/issues/44404
* Resolves: https://github.com/nextcloud/server/issues/45448
* Resolves: https://github.com/nextcloud/server/issues/46935
* Resolves: https://github.com/nextcloud/server/issues/42770
* Resolves: https://github.com/nextcloud/server/issues/34878
* Resolves: https://github.com/nextcloud/server/issues/13696
* Resolves: https://github.com/nextcloud/server/issues/46189
* Closes: #13205

## Summary
- [x] Folder view
- [x] Split of "note-to-recipient" https://github.com/nextcloud/server/pull/46892

## TODO

- [x] needs https://github.com/nextcloud-libraries/nextcloud-upload/pull/1225
- [x] needs a new nextcloud-upload release (1.3.1+)
- [x] needs a new nextcloud-files release (3.5.1+)
- [x] needs a new nextcloud-dialog release (with that files release used)
- [x] File drop
- [x] Single file view

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

susnux committed:
feat(files_sharing): Migrate public share to use Vue files list

Co-authored-by: Ferdinand Thiessen <opensource@fthiessen.de>
Co-authored-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com>
Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(files): Do not use `favorite`, `sidebar` and `view-in-folder` action for public shares

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(files): Fix file previews for public shares

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(files): Adjust files list for sharing permissions on public shares

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

skjnldsv commented on apps/files/src/actions/favoriteAction.ts:
```diff
@@ -68,9 +70,15 @@ export const action = new FileAction({
-		return !nodes.some(node => !node.root?.startsWith?.('/files'))
+	enabled(nodes: Node[], view: View) {
+		// Not enabled for public shares
+		if (view.id === 'public-share') {
+			return false
+		}
```
I would prefer to check if authenticated, seems cleaner

skjnldsv replied:
:wink: 
I saw the WIP, it was a quick overview anyway, no pressure :+1:

skjnldsv commented on apps/files/src/services/RouterService.ts:
```diff
@@ -26,6 +26,14 @@ export default class RouterService {
+	 * This is a protected getter only for internal use
+	 * @private
+	 */
+	get router() {
+		return this._router
+	}
```
Is it really needed? :see_no_evil: 
I'd like to avoid people use it

susnux replied:
To inject the router into the files app, because we need another router for the files_sharing app.
But I will think about an alternative solution

susnux replied:
I did not find any other good solution, we could make it "protected" (call it `_router`) to make it visually private but there is no good other way to inject a different router instance :pensive:

susnux committed:
refactor(files): Make `RouterService` provide a protected getter for the router to allow injecting the `files_sharing` router

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(files): Disable copy-move-action of single file shares

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux commented:
Current blockers:
- [ ] *dialogs* needs an update with updated files and use functions from `@nextcloud/sharing/public`
- [ ] *upload* needs an update

susnux commented:
Missing functions:
* [ ] Note to receiver
* [ ] Add to your cloud

Broken
* [ ] Writeable single file shares currently allow rename
* [ ] Readonly Folder shares offer copy action -> does obviously not work.

susnux committed:
refactor(files_sharing): Make file structure more clear

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux commented:
Status report:
It works and there are only a few todos I like to tackle before I consider this mergeable:
- [ ] Viewer needs to support dav v2 entry point
- [ ] Hide downloads -> We should have a common solution for internal and public links
- [ ] Tests. I really would like to have proper E2E tests (we no have any for the old version but we should take the opportunity to create some to prevent regressions)

And one thing I would consider nice to have to cleanup the code a lot: Provide the current folder to the `enabled` function of a file action. This would allow to just check the current folder instead of using magic values from initial state.

susnux committed:
fix(files): Do not allow rename action on single-file-shares

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
feat(files): Allow to download files on public shares

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(files): Do not allow copy action on public shares without `create` permission

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(files_sharing): Handle download permission the same way for public and internal shares

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(files_sharing): Disable sharing status action for public shares

This is disabled to not leak personal information into the public

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(files_sharing): Set file-request header for nextcloud/upload

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(files): Disable internal views on public shares

This is needed because some views have side effects, like `favorites` doing requests on registration time.

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
chore(files_sharing): Remove now unused legacy scripts

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(files): "New folder" menu entry requires read permission

This will disable "New folder" menu entry for file-drop shares.

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(files_sharing): Ensure sidebar share entry lists are labelled

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
fix(files): Do not add click listener if there is no default action on public shares

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux committed:
test: Add end-to-end tests for new public share Vue UI

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

susnux commented:
When we merge this all apps need to adjust, so `text`, `richdocuments` and `files_pdfviewer` need to handle public shares the same way like they do with `files`.
This cleans up a lot of legacy code, but also meaning those apps are probably broken until they adjust.

susnux committed:
fix(files): "Edit locally" action should not be shown on public shares

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

artonge commented on apps/files_sharing/lib/DefaultPublicShareTemplateProvider.php:
```diff
@@ -175,93 +192,28 @@ public function renderPage(IShare $share, string $token, string $path): Template
-			Util::addScript('files_sharing', 'public_note');
-
-			$download = new SimpleMenuAction('download', $this->l10n->t('Download'), 'icon-download', $shareTmpl['downloadURL'], 0, $shareTmpl['fileSize']);
-			$downloadAll = new SimpleMenuAction('download', $this->l10n->t('Download all files'), 'icon-download', $shareTmpl['downloadURL'], 0, $shareTmpl['fileSize']);
```
Let's keep it or at least ask the designers

artonge reviewed (commented):
- [ ] Check that E2EE file drop still works

artonge commented on apps/files/src/actions/viewInFolderAction.ts:
```diff
@@ -14,6 +15,11 @@ export const action = new FileAction({
 
 	enabled(nodes: Node[], view: View) {
+		// Not enabled for public shares
+		if (isPublicShare()) {
+			return false
+		}
```
Or don't register them? Not sure whether we can do that. Else, this is fine :)

artonge replied:
Whatever

Pytal commented on apps/files_sharing/src/views/FilesViewFileDropEmptyContent.vue:
```diff
@@ -0,0 +1,51 @@
+  - SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+  - SPDX-License-Identifier: AGPL-3.0-or-later
+-->
+<template>
+	<NcEmptyContent :name="t('files_sharing', 'File drop')">
+		<template #icon>
+			<NcIconSvgWrapper :svg="svgCloudUpload" />
+		</template>
+		<template #description>
+			{{ t('files_sharing', 'Upload files to {foldername}.', { foldername }) }}
+			{{ disclaimer === '' ? '' : t('files_sharing', 'By uploading files, you agree to the terms of service.') }}
+		</template>
+		<template v-if="disclaimer" #action>
+			<!-- Terms of service if enabled -->
+			<NcButton type="primary" @click="showDialog = true">
+				{{ t('files_sharing', 'View terms of service') }}
+			</NcButton>
+			<NcDialog close-on-click-outside
+				content-classes="terms-of-service-dialog"
+				:open.sync="showDialog"
+				:name="t('files_sharing', 'Terms of service')"
+				:message="disclaimer" />
+		</template>
+	</NcEmptyContent>
+</template>
```
A "New" button should be added here I think to be consistent with the empty New folder view

come-nc reviewed (approved):
All good on backend side apart from my 2 comments

Pytal commented on apps/files/src/utils/permissions.ts:
```diff
@@ -0,0 +1,29 @@
+/*!
```
Is there documentation somewhere on this comment standard for licenses?

susnux replied:
I do not think there is a standard, it is something that evolved when code minifiers became popular.
By default all comments were remove so someone introduced `/*!` as a **legal comment** so minifiers (e.g. uglifyjs esbuild etc) keep the comments.

I just prefer having those comments because `/**` is always considered a docblock comment, so having this:

```js
/**
 * SPDX
 */
class Foo {
}
```

will always make the SPDX comment part of the `Foo` documentation while it is not a docblock for that class.

susnux committed:
chore: Compile assets

Signed-off-by: Ferdinand Thiessen <opensource@fthiessen.de>

szaimen commented:
🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉

skjnldsv reviewed (commented):
:tada: :tada: :tada: :tada:
